pkg/csi/cinder/controllerserver.go

```diff
@@ -335,73 +381,208 @@ func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
+			return nil, status.Error(codes.Internal, "Multiple backups reported by Cinder with same name")
+		}
+
+		if len(backups) == 1 {
+			backup = &backups[0]
+
+			// Verify the existing backup has the same VolumeID, otherwise it belongs to another volume
+			if backup.VolumeID != volumeID {
+				return nil, status.Error(codes.AlreadyExists, "Backup with given name already exists, with different source volume ID")
+			}
+
+			// If a backup of the volume already exists, skip creating the snapshot
+			backupAlreadyExists = true
+			klog.V(3).Infof("Found existing backup %s from volume with ID: %s", name, volumeID)
+		}
+	}
+
+	// Create the snapshot if the backup does not already exist and wait for it to be ready
+	if !backupAlreadyExists {
+		snap, err = cs.createSnapshot(name, volumeID, req.Parameters)
+		if err != nil {
+			return nil, err
+		}
+
+		ctime = timestamppb.New(snap.CreatedAt)
+		if err = ctime.CheckValid(); err != nil {
+			klog.Errorf("Error to convert time to timestamp: %v", err)
+		}
+
+		err = cs.Cloud.WaitSnapshotReady(snap.ID)
+		if err != nil {
+			klog.Errorf("Failed to WaitSnapshotReady: %v", err)
+			return nil, status.Errorf(codes.Internal, "CreateSnapshot failed with error %v", err)
+		}
+
+		snapSize = snap.Size
+	}
+
+	if snapshotType == "snapshot" {
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SnapshotId:     snap.ID,
+				SizeBytes:      int64(snap.Size * 1024 * 1024 * 1024),
+				SourceVolumeId: snap.VolumeID,
+				CreationTime:   ctime,
+				ReadyToUse:     true,
+			},
+		}, nil
+	}
+
+	// If snapshotType is 'backup', create a backup from the snapshot and delete the snapshot.
+	if snapshotType == "backup" {
+
+		if !backupAlreadyExists {
+			backup, err = cs.createBackup(name, volumeID, snap, req.Parameters)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		ctime = timestamppb.New(backup.CreatedAt)
+		if err := ctime.CheckValid(); err != nil {
+			klog.Errorf("Error to convert time to timestamp: %v", err)
+		}
+
+		// Get the max duration to wait in seconds per GB of snapshot
+		backupMaxDurationSecondsPerGB, err = strconv.Atoi(openstack.BackupMaxDurationPerGB)
+		if err != nil {
+			klog.V(5).Infof("Set backup-max-duration-seconds-per-gb to 20 due to: %v", err)
+			backupMaxDurationSecondsPerGB = openstack.BackupMaxDurationSecondsPerGBDefault
+		}
+
+		err = cs.Cloud.WaitBackupReady(backup.ID, snapSize, backupMaxDurationSecondsPerGB)
+		if err != nil {
+			klog.Errorf("Failed to WaitBackupReady: %v", err)
+			return nil, status.Error(codes.Internal, fmt.Sprintf("CreateBackup failed with error %v", err))
+		}
+
+		// Necessary to get all the backup information, including size.
+		backup, err = cs.Cloud.GetBackupByID(backup.ID)
+		if err != nil {
+			klog.Errorf("Failed to GetBackupByID after backup creation: %v", err)
+			return nil, status.Error(codes.Internal, fmt.Sprintf("GetBackupByID failed with error %v", err))
+		}
+
+		if !backupAlreadyExists {
```

Comment:
What if backup and snapshot got created, but then `GetBackupByID()` in line 480 failed because of some transient reason? That way we orphan the snapshot. I think we should always attempt to delete the snapshot if `snapshotType="backup"`.

Reply:
Agree! Same with previous comment, could you help me with the method to match IsNotFound errors so I can ignore them?

Reply:
[This should be it, isn't it? ](https://github.com/kubernetes/cloud-provider-openstack/blob/1fdebaa1c3c29fd27772c75c2a0d9b512cdd7983/pkg/util/errors/errors.go#L45-L65)

Your code is using it in a few places already.